File: Array/27. Remove Element/solution 2/27.c
// 雙指針法（快慢指針法）：透過一個快指針和慢指針在一個循環完成下兩個循環的工作。

/* 定義快慢指針

    快指標：尋找新的儲存元素，新的儲存就是不含目標元素的存儲
    慢標指標：指向更新新記憶體下的位置
*/

// Time Complexity：O(n)

// Space Complexity：O(1)

int removeElement(int* nums, int numsSize, int val) {
    int slow = 0;
    for(int fast=0; fast<numsSize; fast++){
        if(val != nums[fast]){
            nums[slow++] = nums[fast];
        }
    }
    return slow;
}

File: Array/27. Remove Element/solution 1/27.java
// 暴力破解法

// Time Complexity： O(n^2)

// Space Complexity：O(1)

class Solution {
    public int removeElement(int[] nums, int val) {
        int size = nums.length;

        for(int i=0; i<size; i++){
            if(nums[i] == val){
                for(int j=i+1; j<size; j++){
                    nums[j-1]=nums[j];
                }
                i--;
                size--;
            }
        }
        return size;
    }
}
